src/qibo/models/circuit.py

```diff
@@ -575,7 +574,15 @@ def add(self, gate):
                     )
 
             if isinstance(gate, gates.M):
-                self.add(gate.basis)
+                # The following loop is useful when two circuits are added together:
+                # all the gates in the basis of the measure gates should not
+                # be added to the new circuit, otherwise once the measure gate is added in the circuit
+                # there will be two of the same.
+
+                for base in gate.basis:
+                    if base not in self.queue:
```

Comment:
Is `base` compared by some sort of ID? Because I believe `self.queue` to be just a list and not implementing explicitly `__contains__`:
https://github.com/qiboteam/qibo/blob/87ed54f203db477af604429d8a898e41b96d5056/src/qibo/models/circuit.py#L32

So, I believe `base` won't be added *if and only if* exactly the same object is not already part of `self.queue`. If this is fine, that's the correct way, otherwise we need to find a better comparison.